akka-stream-typed/src/main/scala/akka/stream/typed/scaladsl/ActorFlow.scala

```diff
@@ -0,0 +1,107 @@
+import akka.stream._
+import akka.util.{ ConstantFun, Timeout }
+
+import scala.annotation.implicitNotFound
+import scala.concurrent.{ ExecutionContextExecutor, Future }
+import scala.reflect.ClassTag
+
+/**
+ * Collection of Sinks aimed at integrating with typed Actors.
+ */
+object ActorFlow {
+
+  implicit class ActorFlowOps[I, O, M](val flow: Flow[I, O, M]) extends AnyVal {
+
+  }
+
+  /**
+   * Use the `ask` pattern to send a request-reply message to the target `ref` actor.
+   * If any of the asks times out it will fail the stream with a [[akka.pattern.AskTimeoutException]].
+   *
+   * Do not forget to include the expected response type in the method call, like so:
+   *
+   * {{{
+   * flow.ask[ExpectedReply](ref, (el, replyTo) => Asking(el, replyTo))
+   * // or
+   * flow.ask[ExpectedReply](ref, (el, replyTo) => el.replyTo(replyTo))
+   * flow.ask[ExpectedReply](ref, _.replyTo(_)) // FIXME does this work
+   * }}}
+   *
+   * otherwise `Nothing` will be assumed, which is most likely not what you want.
+   *
+   * Defaults to parallelism of 2 messages in flight, since while one ask message may be being worked on, the second one
+   * still be in the mailbox, so defaulting to sending the second one a bit earlier than when first ask has replied maintains
+   * a slightly healthier throughput.
+   *
+   * The mapTo class parameter is used to cast the incoming responses to the expected response type.
+   *
+   * Similar to the plain ask pattern, the target actor is allowed to reply with `akka.util.Status`.
+   * An `akka.util.Status#Failure` will cause the stage to fail with the cause carried in the `Failure` message.
+   *
+   * The stage fails with an [[akka.stream.WatchedActorTerminatedException]] if the target actor is terminated.
+   *
+   * Adheres to the [[ActorAttributes.SupervisionStrategy]] attribute.
+   *
+   * '''Emits when''' the futures (in submission order) created by the ask pattern internally are completed
+   *
+   * '''Backpressures when''' the number of futures reaches the configured parallelism and the downstream backpressures
+   *
+   * '''Completes when''' upstream completes and all futures have been completed and all elements have been emitted
+   *
+   * '''Fails when''' the passed in actor terminates, or a timeout is exceeded in any of the asks performed
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  @implicitNotFound("Missing an implicit akka.util.Timeout for the ask() stage")
+  def ask[Q, O, O2, Mat](ref: ActorRef[Q])(f: (O, ActorRef[O2]) ⇒ Q)(implicit timeout: Timeout, tag: ClassTag[O2]): Flow[O, O2, NotUsed] = {
+    // such implementation would need implicit scheduler
+    //    flow.mapAsync[O2](2) { (el: O) ⇒
+    //      import akka.actor.typed.scaladsl.AskPattern._
+    //      // ref ?[O2] (replyTo ⇒ f(el, replyTo)) // <<<< this needs scheduler
+    //    }
+
+    Flow[O].via(new TypedAskStage(ref, f))
+      .mapAsync(2)(ConstantFun.scalaIdentityFunction)
+      .named("ask")
+  }
```

Comment:
Since we can convert typed ActorRef to untyped, and vice versa, we might be able to implement this with the untyped ask and watch in the same way as in the untyped FlowOps.ask, apart from that we would use the explicit ask with the `messageFactory: ActorRef ⇒ Any`?

That would also cover watch.

Reply:
Yeah, that could work I guess... since we guaranteed type-correctness on the type signature level converting and using the untyped may work indeed... Will give it a try.

Reply:
another minor detail is if we would like to convert the AskTimeoutException to a j.u.c.TimeoutException, which is what I think we use in Typed. Should be easy to recovery the Future.

Reply:
Thanks for the hint!